Source/JavaScriptCore/wasm/WasmTypeDefinition.h

```diff
@@ -911,8 +912,9 @@ class TypeInformation {
     static void tryCleanup();
 private:
     HashSet<Wasm::TypeHash> m_typeSet;
-    HashMap<TypeIndex, TypeIndex> m_unrollingCache;
+    HashMap<TypeIndex, RefPtr<const TypeDefinition>> m_unrollingCache;
```

Comment:
The fact that these need to be globally held is kinda terrible. I guess the spec forces this though, since any two separately compiled wasm gc modules need to play nicely together. 

I am worried because we don't have a plan to collect these once a wasm module is destroyed. We have this problem for wasm jit code as well, but in this case it seems much harder to solve. 

Perhaps in the future, we should make sure that we can handle null here

Reply:
For this specific thing (the `m_unrollingCache`) it is optional that we hold onto it, in that this is a perf optimization and not here for correctness. It's just here to avoid doing type expansion every time it's needed.

And while the values are held strongly, the keys are not. So if when `tryCleanup` is called when a wasm module is destroyed the refcount is 1, the key will actually be removed (in [this code](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp#L1106)) and the whole mapping collected.

If you mean the other thing I added for the placeholders, yes you are right that they have to be held globally. I think it might be possible to try to clean these up too if no live wasm modules are using recursive types, but it's not trivial to compute this condition I think.